### DISCUSSION TITLE
error: size of sigstruct size does not match

### DISCUSSION THREAD

**Discussion by pdavis39** [gramineproject/gramine#1997] [General]
I have several issues with a python application I am trying to run

when I invoke make SGX=1 

I get the warning message below that I have hightlighted and do not understand how to resolve the warning.

gramine-manifest \
                -DARCH_LIBDIR=/lib/x86_64-linux-gnu \
                -DAPPDIR=/nfs/enclave_a/demo/mstar \
                -DPYTHONDIR=/usr/local \
                -DGRAMINEDIR=/usr/local/gramine \
                -DMODELDIR=/path/to/model \
                -DLOG_LEVEL=error \
                sgx_ztee_mstar_classifier_final.manifest.template > sgx_ztee_mstar_classifier_final.manifest.sgx
**WARNING: error in manifest (after rendering): extra keys not allowed @ data['loader']['args']**
gramine-sgx-sign \
        --key /XXXX/XXXX/.config/gramine/enclave-key.pem \
                --manifest sgx_ztee_mstar_classifier_final.manifest.sgx \
                --output sgx_ztee_mstar_classifier_final.sig

When I run 

gramine-sgx sgx_ztee_mstar_classifier_final

Gramine is starting. Parsing TOML manifest file, this may take some time...
**error: size of sigstruct size does not match
error: Reading enclave sigstruct failed: Invalid argument (EINVAL)
error: load_enclave() failed with error: Invalid argument (EINVAL)**

It is not clear where I need to resolve this error.  

Thank in advance for support

**Comment by pdavis39** [comment]
did some more work 

We are accessing the data through an NFS mount.
  

 SGX=1 DEBUG=1
make: Warning: File 'sgx_ztee_mstar_classifier_final.py' has modification time 572 s in the future
gramine-manifest \
        -DARCH_LIBDIR=/lib/x86_64-linux-gnu \
        -DAPPDIR=/nfs/enclave_a/demo/mstar \
        -DPYTHONDIR=/usr/local \
        -DGRAMINEDIR=/usr/local/gramine \
        -DMODELDIR=/path/to/model \
        -DLOG_LEVEL=debug \
        sgx_ztee_mstar_classifier_final.manifest.template > sgx_ztee_mstar_classifier_final.manifest.sgx
**WARNING: error in manifest (after rendering): extra keys not allowed @ data['loader']['exec']**
gramine-sgx-sign \
                --key /xxxxx/xxxxx/.config/gramine/enclave-key.pem \
                --manifest sgx_ztee_mstar_classifier_final.manifest.sgx \
                --output sgx_ztee_mstar_classifier_final.sig
Attributes (required for enclave measurement):
    size:        0x40000000
    edmm:        False
    max_threads: 4
SGX remote attestation:
    None
Memory:
    000000003ffff000-0000000040000000 [REG:R--] (manifest) measured
    000000003ffdf000-000000003ffff000 [REG:RW-] (ssa) measured
    000000003ffdb000-000000003ffdf000 [TCS:---] (tcs) measured
    000000003ffd7000-000000003ffdb000 [REG:RW-] (tls) measured
    000000003ff97000-000000003ffd7000 [REG:RW-] (stack) measured
    000000003ff57000-000000003ff97000 [REG:RW-] (stack) measured
    000000003ff17000-000000003ff57000 [REG:RW-] (stack) measured
    000000003fed7000-000000003ff17000 [REG:RW-] (stack) measured
    000000003fec7000-000000003fed7000 [REG:RW-] (sig_stack) measured
    000000003feb7000-000000003fec7000 [REG:RW-] (sig_stack) measured
    000000003fea7000-000000003feb7000 [REG:RW-] (sig_stack) measured
    000000003fe97000-000000003fea7000 [REG:RW-] (sig_stack) measured
    000000003fe3d000-000000003fe8d000 [REG:R-X] (code) measured
    000000003fe8d000-000000003fe97000 [REG:RW-] (data) measured
    0000000000010000-000000003fe3d000 [REG:RWX] (free)
Measurement:
    fb798c5bfb3fbf27ee5619ad5db048ffa82e9d6e4b9c976ae1f722b3f8715da4
**make: warning:  Clock skew detected.  Your build may be incomplete.** 


gramine-sgx sgx_ztee_mstar_classifier_final
Gramine is starting. Parsing TOML manifest file, this may take some time...
**error: size of sigstruct size does not match
error: Reading enclave sigstruct failed: Invalid argument (EINVAL)
error: load_enclave() failed with error: Invalid argument (EINVAL)**

Looking forward to thoughts on how to fix this

**Comment by woju** [comment]
(please wrap your logs in triple-backtics, \`\`\`)

```
gramine-sgx-sign \
                --key /xxxxx/xxxxx/.config/gramine/enclave-key.pem \
                --manifest sgx_ztee_mstar_classifier_final.manifest.sgx \
                --output sgx_ztee_mstar_classifier_final.sig
```
So this `sgx_ztee_mstar_classifier_final.sig` file that was supposedly created by this command should be exactly 1808 bytes long. Is it really?

Cf. `CI-Examples/helloworld`:
```
% ls -l
total 56
-rwxr-xr-x 1 woju woju 15960 Aug 29 17:13 helloworld
-rw-r--r-- 1 woju woju   171 Aug 29 17:10 helloworld.c
-rw-r--r-- 1 woju woju   810 Aug 29 17:13 helloworld.manifest
-rw-r--r-- 1 woju woju 10261 Aug 29 17:13 helloworld.manifest.sgx
-rw-r--r-- 1 woju woju   646 Aug 29 17:13 helloworld.manifest.template
-rw-r--r-- 1 woju woju  1448 Aug 29 17:13 helloworld.o
-rw-r--r-- 1 woju woju  1808 Aug 29 17:13 helloworld.sig
-rw-r--r-- 1 woju woju  1991 Aug 29 17:10 Makefile
-rw-r--r-- 1 woju woju     0 Aug 29 17:13 OUTPUT
-rw-r--r-- 1 woju woju   522 Aug 29 17:10 README.md
```
(note `helloworld.sig` size; other files might have different sizes, but this one needs exact size)

**Comment by mkow** [comment]
Also, please fix your manifest:

> WARNING: error in manifest (after rendering): extra keys not allowed @ data['loader']['args']

I don't know what you have there, but it seems it doesn't conform to our schema.

**Comment by dimakuv** [comment]
In case you want to know where this error message comes from: https://github.com/gramineproject/gramine/blob/bc479f1f81e57978222a9ce1e80d01cec6fb8b00/pal/src/host/linux-sgx/host_framework.c#L139-L142

So clearly some issue with this `sgx_ztee_mstar_classifier_final.sig` file (in particular, its size).

**Comment by pdavis39** [comment]
I've gotten past my last issue working on deploying a Python convolutional neural network (CNN) application within an Intel SGX enclave using Gramine. The application uses common machine learning libraries such as TensorFlow, Keras, and OpenCV.

Despite making progress, I'm currently facing issues related to the application's dependencies and the ELF binary loading process. The error messages I'm receiving are as follows:

**Gramine is starting. Parsing TOML manifest file, this may take some time...
error: Loaded binary has more than one dependency (must be only one dependency -- the PAL binary)
error: Could not map the ELF file into memory and then relocate it
error: PAL failed Unable to load loader.entrypoint: -4**

Here are some relevant details from my setup:

Makefile Variables:

ENTRYPOINT = /usr/bin/python3.9
BASE_DIRECTORY = /nfs//demo/mstar
LIBCDIR = /lib64/
APPDIR = /nfs/demo/mstar
PYTHONDIR = /usr/lib64/python3.9
GRAMINEDIR = /usr/bin/gramine-sgx
MODELDIR = /nfs/model

Manifest Template Snippet:

loader.entrypoint = "file:/usr/bin/python3.9"
# ... environment variables and mount points ...
sgx.enclave_size = "32G"
sgx.thread_num = "8"

# Define mount points
fs.mount.lib.type = "chroot"
fs.mount.lib.path = "/lib"
fs.mount.lib.uri = "file:$(LIBCDIR)"

fs.mount.usr.path = "/usr"
fs.mount.usr.uri = "file:$(USRLIBDIR)"

fs.mount.etc.type = "chroot"
fs.mount.etc.path = "/etc"
fs.mount.etc.uri = "file:$(ETCDIR)"


fs.mount.app.type = "chroot"
fs.mount.app.path = "/app"
fs.mount.app.uri = "file:/nfs//demo/mstar"

# ... trusted files ...
sgx.trusted_files = [
        { uri = "file:/nfs//model/mstar_CNN3.weights.h5" },
        { uri = "file:/nfs//demo/mstar/sgx_ztee_mstar_classifier_final.py" },
        { uri = "file:{{ base_directory }}/data/MSTAR-8-Classes/" },
        { uri = "file:{{ entrypoint }}" },
        { uri = "file:/lib64/libpython3.9.so.1.0" },
        { uri = "file:/lib64/libc.so.6" },
        { uri = "file:/lib64/libm.so.6" },
        { uri = "file:/lib64/ld-linux-x86-64.so.2" },
        { uri = "file:/usr/lib64/python3.9/" },
        { uri = "file:/usr/lib64/python3.9/lib-dynload/" },
        { uri = "file:/XXXX/XXXX/.local/lib/python3.9/site-packages/" },
        { uri = "file:/usr/lib64/python3.9/site-packages/" },
        { uri = "file:/usr/lib/python3.9/site-packages/" },

I've included the necessary shared libraries in the sgx.trusted_files and set up the fs.mount points to match the expected file system structure. However, the errors persist, indicating an issue with the binary dependencies and loading.

I would greatly appreciate any guidance or insights from those who have experience with Gramine and Intel SGX, particularly in the context of Python applications with complex dependencies.

**Comment by pdavis39** [comment]
think I need to try a different approach thanks for the help
